taichi/transforms/determine_ad_stack_size.cpp

```diff
@@ -0,0 +1,34 @@
+#include "taichi/ir/analysis.h"
+#include "taichi/ir/control_flow_graph.h"
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/transforms.h"
+
+#include <queue>
+#include <unordered_map>
+
+namespace taichi {
+namespace lang {
+
+namespace irpass {
+
+bool determine_ad_stack_size(IRNode *root, const CompileConfig &config) {
+  if (irpass::analysis::gather_statements(root, [&](Stmt *s) {
+        if (auto ad_stack = s->cast<AdStackAllocaStmt>()) {
+          return ad_stack->max_size == 0;  // adaptive
+        }
+        return false;
+      }).empty()) {
+    return false;  // no AD-stacks with adaptive size
+  }
+  auto cfg = analysis::build_cfg(root);
+  cfg->simplify_graph();
+  constexpr int kMaxAdStackSize = 32;
+  cfg->determine_ad_stack_size(kMaxAdStackSize);
```

Comment:
Instead of using `32` here, maybe we should add an option `max_ad_stack_size` to `CompileConfig`. Note that `CompileConfig::ad_stack_size` was used as the default stack size when no size determination mechanism is used.